### PR TITLE
Dedicated table for external (OIDC) logins

### PR DESCRIPTION
**Pull request by dmitrys-odysseus**

### CODE REVIEW

File: src/main/java/com/odysseusinc/arachne/portal/model/ExternalLogin.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    protected String provider;
+
+    @Column(nullable = false, name = "sub")
+    protected String sub;
+
+    @Column(name = "email")
+    protected String email;
+
+    @Column(name = "created")
+    private Instant created;
```
Comment (YaroslavTir) [inline src/main/java/com/odysseusinc/arachne/portal/model/ExternalLogin.java:52]: Have you discussed using `Instant` with others?

File: src/main/java/com/odysseusinc/arachne/portal/model/ExternalLogin.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+    @ManyToOne(targetEntity = User.class, cascade = CascadeType.PERSIST)
```
Comment (YaroslavTir) [inline src/main/java/com/odysseusinc/arachne/portal/model/ExternalLogin.java:54]: do you really need CascadeType.PERSIST here?

File: src/main/java/com/odysseusinc/arachne/portal/model/factory/ArachneUserFactory.java [hunk 3]
```diff
@@ -53,4 +53,13 @@ public static ArachneUser create(IUser user) {
         );
     }
 
+    public static ArachneUser createExternalDummy(ExternalLogin login) {
+        String email = login.getEmail();
+        // TODO should it have any roles?
+        List<GrantedAuthority> authorities = Collections.emptyList();
+        // TODO something better here
+        long id = -login.getId();
```
Comment (YaroslavTir) [inline src/main/java/com/odysseusinc/arachne/portal/model/factory/ArachneUserFactory.java:61]: hmm, that looks like magic here. Can you add an explanation here?

File: src/main/java/com/odysseusinc/arachne/portal/security/Oauth2SuccessHandler.java [hunk 4]
```diff
@@ -64,16 +72,20 @@ public void onAuthenticationSuccess(
...
-            ).map(user -> {
-                return tokenProvider.createToken(user.getUsername(), additionalInfo, null);
-            }).orElseGet(() -> {
-                ArachneUser details = createUser(attributes, username, method);
-                return tokenProvider.createToken(details.getUsername(), additionalInfo, null);
-            });
+            String sub = (String) attributes.get(StandardClaimNames.SUB);
+            String email = (String) attributes.get(StandardClaimNames.EMAIL);
+            String issuer = Optional.ofNullable((String) attributes.get(IdTokenClaimNames.ISS)).orElseGet(() ->
+                    // Fall back to use domain from sub, as some providers don't send issuer
```
Comment (YaroslavTir) [inline src/main/java/com/odysseusinc/arachne/portal/security/Oauth2SuccessHandler.java:78]: *a/the* domain from sub

File: src/main/java/com/odysseusinc/arachne/portal/service/ExternalLoginService.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+     */
+    @Value("${user.external.autolink:false}")
+    private boolean autolink;
+
+    /**
+     * Specifies whether new external users have their profile imported from authentication provider.
+     * Note that if autolink is enabled, it will be attempted first.
+     */
+    @Value("${user.external.autoimport:true}")
+    private boolean autoimport;
```
Comment (YaroslavTir) [inline src/main/java/com/odysseusinc/arachne/portal/service/ExternalLoginService.java:37]: If the flag is not used, it is better to remove it.

File: src/main/java/com/odysseusinc/arachne/portal/service/ExternalLoginService.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+
+@Service
+public class ExternalLoginService {
+    private static final Logger log = LoggerFactory.getLogger(ExternalLoginService.class);
```
[the remark this answers is not in the document]
Comment (dmitrys-odysseus) [reply src/main/java/com/odysseusinc/arachne/portal/service/ExternalLoginService.java:24]: Would be nice to have, however lombok is not on board on this project for the moment.

File: src/main/java/com/odysseusinc/arachne/portal/service/ExternalLoginService.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    private EntityManager em;
+
+    @Autowired
+    private UserRepository userRepo;
+
+    @Transactional
+    public <U extends IUser> ExternalLogin login(String provider, String sub, String email, Supplier<U> createUser) {
+        CriteriaBuilder cb = em.getCriteriaBuilder();
+        CriteriaQuery<ExternalLogin> q = cb.createQuery(ExternalLogin.class);
+        Root<ExternalLogin> root = q.from(ExternalLogin.class);
```
[the remark this answers is not in the document]
Comment (dmitrys-odysseus) [reply src/main/java/com/odysseusinc/arachne/portal/service/ExternalLoginService.java:49]: Spring-data repositories are one big tightly coupled ball, when they start to depend on each other. 
Strategically, we need clean generated metamodels, however both EntityGraph and especially entity inheritance in some cases make this movement quite hard. However, we shouldn't make this direction of evolution any harder, at least.

File: src/main/java/com/odysseusinc/arachne/portal/model/ExternalLogin.java (hunk 1 above)
Comment (dmitrys-odysseus) [reply src/main/java/com/odysseusinc/arachne/portal/model/ExternalLogin.java:52]: Other places in code use "new Date()", which is locale-sensitive. Not sure what you are proposing here, replicate those issues?

File: src/main/java/com/odysseusinc/arachne/portal/model/ExternalLogin.java (hunk 2 above)
Comment (dmitrys-odysseus) [reply src/main/java/com/odysseusinc/arachne/portal/model/ExternalLogin.java:54]: For the moment - probably no.

File: src/main/java/com/odysseusinc/arachne/portal/model/factory/ArachneUserFactory.java (hunk 3 above)
Comment (dmitrys-odysseus) [reply src/main/java/com/odysseusinc/arachne/portal/model/factory/ArachneUserFactory.java:61]: Method removed

File: src/main/java/com/odysseusinc/arachne/portal/service/ExternalLoginService.java (hunk 5 above)
Comment (dmitrys-odysseus) [reply src/main/java/com/odysseusinc/arachne/portal/service/ExternalLoginService.java]: Removed

**Review by YaroslavTir — APPROVED** [review]

**Pull request merged**